app/models/source_resolver.rb

```diff
@@ -1,34 +1,62 @@
 class SourceResolver
-  def self.resolve(new_source)
-    new(new_source).resolve
+  def self.resolve(adjusted_source)
+    new(adjusted_source).resolve
   end
 
-  def initialize(new_source)
-    @new_source = new_source
+  def initialize(adjusted_source)
+    @adjusted_source = adjusted_source
   end
 
   def resolve
-    return if @new_source.save
+    return if @adjusted_source.save
     adjust_ranks
-    @new_source.save
+    @adjusted_source.save
   end
 
   private
 
+  def conflicted_rank_types
+    @adjusted_source.errors.keys
+  end
+
   def adjust_ranks
     conflicted_rank_types.each do |rank_type|
-      sources_to_adjust(rank_type).each do |source|
-        source.increment!(rank_type)
+      if demotion?(rank_type)
+        move_sources_up(rank_type)
+      else
+        move_sources_down(rank_type)
       end
     end
   end
 
-  def conflicted_rank_types
-    @new_source.errors.keys
+  def demotion?(rank_type)
+    return false unless @adjusted_source.persisted?
+
+    old, new = @adjusted_source.changes[rank_type]
+    old < new
+  end
+
+  def move_sources_up(rank_type)
+    sources_to_decrement(rank_type).each do |source|
+      source.decrement!(rank_type)
+    end
+  end
+
+  def sources_to_decrement(rank_type)
+    cutoff = @adjusted_source.rank_for rank_type
+    Source.where("#{rank_type} <= ?", cutoff).order(rank_type)
+  end
+
+  def move_sources_down(rank_type)
+    sources_to_increment(rank_type).each do |source|
+      source.increment!(rank_type)
+    end
   end
 
-  def sources_to_adjust(rank_type)
-    cutoff = @new_source.rank_for rank_type
-    Source.where("#{rank_type} >= ?", cutoff).order("#{rank_type} DESC")
+  def sources_to_increment(rank_type)
+    cutoff = @adjusted_source.rank_for rank_type
+    old_rank = @adjusted_source.changes[rank_type].first || Source.count + 1
```

Comment:
In the case where `adjusted_source` is a new source - check it:

```
> existing_source = Source.first
> existing_source.email_rank = 2
=> 2
> existing_source.changes
=> {"email_rank"=>[0, 2]}
> new_source = Source.create email_rank: 3
> new_source.changes
=> {"email_rank"=>[nil, 3]}
```

So, I had to have a fallback for the case where you're adding a new source. In that case you want to move down all sources that follow the rank of the new one. I could have added a third type of `move_sources_BLAH` method, but since this was the only difference, I kept it with the other.

Maybe it's worth pulling out a couple local variables here to name these concepts a little better?